### PR TITLE
[TR] PIM-5008: Optimize PEF when dealing with lots of attributes

### PR DESCRIPTION
**Pull request by phaseinducer**
https://akeneo.atlassian.net/browse/PIM-5008

| Q                 | A
| ----------------- | ---
| Specs             | no regression
| Behats            | n/a
| Blue CI           | yes
| Changelog updated | yes
| Review and 2 GTM  |

What this PR does ?
- Each time we used to fetch all attributes, fetch only needed attributes by codes instead.
- Rework the "add attributes" extension (the last one that really needs all attributes) so that the attributes choices are fetched only when the user opens the multiselect.
- Load the associations grid only when the user opens the associations tab.
- Fix the way identifier attribute is fetched.

### CODE REVIEW

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/fetcher/attribute-fetcher.js [hunk 1]
```diff
@@ -1,11 +1,51 @@
...
-            }).promise();
+        identifierPromise: null,
+
+        /**
+         * Return the identifier attribute
+         *
+         * @returns {Promise}
+         */
+        getIdentifierAttribute: function () {
+            if (!_.isNull(this.identifierPromise)) {
```
Comment (filipsalpe) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/fetcher/attribute-fetcher.js]:
```
_.isNull = function(obj) {
    return obj === null;
};
```

Seems quite reduntant to rely on underscore for checking if a variable is null :)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/fetcher/attribute-fetcher.js [hunk 2]
```diff
@@ -1,11 +1,51 @@
...
+
+            return this.fetchByTypes([this.options.identifier_type])
+                .then(function (attributes) {
+                    if (0 < attributes.length) {
+                        this.identifierPromise = $.Deferred().resolve(attributes[0]);
+
+                        return this.identifierPromise;
+                    }
+
+                    return $.Deferred().resolve(null);
```
Comment (filipsalpe) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/fetcher/attribute-fetcher.js]: Why not reject in this case?

Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/fetcher/attribute-fetcher.js]: That's right ! Thanks @filipsalpe =)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/fetcher/attribute-fetcher.js [hunk 3]
```diff
@@ -1,11 +1,53 @@
...
-            }).promise();
+        identifierPromise: null,
+
+        /**
+         * Return the identifier attribute
+         *
+         * @return {Promise}
+         */
+        getIdentifierAttribute: function () {
+            if (null !== this.identifierPromise) {
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/fetcher/attribute-fetcher.js]: please turn it like in other fetchers:
```
if (null === this.identifierPromise) {

}

return this.identifierPromise;
```

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/AttributeController.php [hunk 4]
```diff
@@ -45,10 +46,18 @@ public function __construct(
...
+        $criteria = [];
+        if ($request->query->has('identifiers')) {
+            $criteria['code'] = explode(',', $request->query->get('identifiers'));
+        }
+        if ($request->query->has('types')) {
+            $criteria['attributeType'] = explode(',', $request->query->get('types'));
+        }
+
+        $attributes         = $this->attributeRepository->findBy($criteria);
+        $filteredAttributes = $this->collectionFilter
```
Comment (ClementGautier) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/AttributeController.php]: Do we still need to filter the attributes ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/fetcher/attribute-fetcher.js [hunk 5]
```diff
@@ -1,11 +1,53 @@
...
+                return this.identifierPromise;
+            }
+
+            return this.fetchByTypes([this.options.identifier_type])
+                .then(function (attributes) {
+                    if (0 < attributes.length) {
```
Comment (jmleroux) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/fetcher/attribute-fetcher.js]: yoda style is not mandatory for this kind of comparison

**Comment by sumbobyboys** [comment]
GTM after comment fix. Awesome work dude, the pef will be really faster with this improvements

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/AttributeController.php (hunk 4 above)
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/AttributeController.php]: For permissions on attribute groups yes, why ?

**Comment by jmleroux** [comment]
GTM

**Pull request merged**